beets/metadata_plugins.py

```diff
@@ -0,0 +1,364 @@
+
+    @abc.abstractmethod
+    def track_for_id(self, track_id: str) -> TrackInfo | None:
+        """Return a :py:class:`AlbumInfo` object or None if no matching release was
+        found.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def candidates(
+        self,
+        items: Sequence[Item],
+        artist: str,
+        album: str,
+        va_likely: bool,
+    ) -> Iterable[AlbumInfo]:
+        """Return :py:class:`AlbumInfo` candidates that match the given album.
+
+        Used in the autotag functionality to search for albums.
+
+        :param items: List of items in the album
+        :param artist: Album artist
+        :param album: Album name
+        :param va_likely: Whether the album is likely to be by various artists
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def item_candidates(
+        self, item: Item, artist: str, title: str
+    ) -> Iterable[TrackInfo]:
+        """Return :py:class:`TrackInfo` candidates that match the given track.
+
+        Used in the autotag functionality to search for tracks.
+
+        :param item: Track item
+        :param artist: Track artist
+        :param title: Track title
+        """
+        raise NotImplementedError
+
+    def albums_for_ids(self, ids: Sequence[str]) -> Iterable[AlbumInfo | None]:
+        """Batch lookup of album metadata for a list of album IDs.
+
+        Given a list of album identifiers, yields corresponding AlbumInfo
+        objects. Missing albums result in None values in the output iterator.
+        Plugins may implement this for optimized batched lookups instead of
+        single calls to album_for_id.
+        """
+
+        return (self.album_for_id(id) for id in ids)
+
+    def tracks_for_ids(self, ids: Sequence[str]) -> Iterable[TrackInfo | None]:
+        """Batch lookup of track metadata for a list of track IDs.
+
+        Given a list of track identifiers, yields corresponding TrackInfo objects.
+        Missing tracks result in None values in the output iterator. Plugins may
+        implement this for optimized batched lookups instead of single calls to
+        track_for_id.
+        """
+
+        return (self.track_for_id(id) for id in ids)
+
+    def album_distance(
+        self,
+        items: Sequence[Item],
+        album_info: AlbumInfo,
+        mapping: dict[Item, TrackInfo],
+    ) -> Distance:
+        return _get_distance(
+            data_source=self.data_source, info=album_info, config=self.config
+        )
+
+    def track_distance(
+        self,
+        item: Item,
+        info: TrackInfo,
+    ) -> Distance:
+        return _get_distance(
+            data_source=self.data_source, info=info, config=self.config
+        )
+
+    @cached_classproperty
+    def data_source(cls) -> str:
+        """The data source name for this plugin.
+
+        This is inferred from the plugin name.
+        """
+        return cls.__name__.replace("Plugin", "")
```

Comment:
![image](https://github.com/user-attachments/assets/4088e851-7baf-4575-b0f8-dcb9d482d562)
I saw the exact same issue when I implemented `beetsplug.lyrics.Backend` class. I attempted to add types to `cached_classproperty` but failed and ended up using a metaclass:

```py
class BackendClass(type):
    @property
    def name(cls) -> str:
        """Return lowercase name of the backend class."""
        return cls.__name__.lower()


class Backend(RequestHandler, metaclass=BackendClass):
    def __init__(self, config, log):
        self._log = log
        self.config = config

    def fetch(
        self, artist: str, title: str, album: str, length: int
    ) -> tuple[str, str] | None:
        raise NotImplementedError
```

Reply:
Maybe we could instead just add `type: ignore[type-defined]`?

Technically, __it is__ a class attribute, and this error comes up because `cached_classproperty` is not properly typed, so we can safely ignore this, I think.